### ISSUE TITLE
[HOLD for payment 2024-02-22] [$500]  iOS - Login - The splash screen glitches and flashes on launch

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#34696]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.26.1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Open the iOS app
2. Kill the app and launch it again

## Expected Result:
The splash screen transition should be smooth on launch

## Actual Result:
The splash screen glitches and flashes on launch

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/28439d0f-0ac8-4a22-816e-5b2b0c1a7071

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01782ced06468ec4f5</li>
        <li>Upwork Job ID: 1747794137711603712</li>
        <li>Last Price Increase: 2024-01-18</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Reviewer | 28121977</li>
<li>ikevin127 | Contributor | 28121978</li>
</ul></li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
[#wave9-collect-signup](https://expensify.enterprise.slack.com/archives/C05NJ4SLBMF)

**Comment by ikevin127** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

The splash screen hide animation glitches on app launch.

## What is the root cause of that problem?

PR https://github.com/Expensify/App/pull/30905 implemented `expo-image` for rendering of the SVG background image which made the loading process heavier than before when we used `react-native-fast-image`.

On this issue there are 2 things on the `SignInPageLayout` which cause the splash screen hide animation glitch - I will list them in order of severity:
1. The `BackgroundImage` which on iOS `.../BackgroundImage/index.ios.js` uses Image from `expo-image` to load the background image - a large SVG file of ~800kb.

<details>
<summary>home-background--mobile-new.svg</summary>

![home-background--mobile-new](https://github.com/Expensify/App/assets/56457735/0ce7645c-5318-48b4-af21-416a4e5e7ef1)

</details>

2. The `.../SignInPageContent/SignInHeroImage.js` which is the Lottie `Hands` animation.

Both of these are being rendered while the `.../SplashScreenHider/index.native.tsx` component's onLayout calls the splash screen hide function and the callback animates out 2 properties (opacity and scale) of the splash screen background and expensify logo.

## What changes do you think we should make in order to solve the problem?

**TLDR:** Use `isSplashHidden` from SplashScreenHiddenContext to prevent the rendering of both iOS `BackgroundImage` and `SignInHeroImage` Lottie animation until the splash screen is hidden.

Pseudo code:

1. For `.../BackgroundImage/index.ios.js` we will show it with a smooth 750ms animation as soon as the splash hides:
- check if !isSplashHidden is true and return; in order to prevent the Image rendering (which causes the issue)
- below, wrap the Image with Reanimated.View, passing the same styles as Image + animatedStyle
- add `onLoadEnd` prop to the Image where we pass the function that will trigger the animation

this way when `isSplashHidden` becomes true and the SVG finished loading, it will animate-in nicely.

2. For `.../SplashScreenHider/index.native.tsx` we will render an empty View with the Lottie component's styles so that the layout doesn't jump when the actual Lottie component is rendered:
- check if !isSplashHidden is true and return the empty View with the Lottie component's styles - by doing this we prevent the rendering of the Lottie component (which causes the issue)

this way when `isSplashHidden` becomes true only then we render the Lottie component that will display the Hands animation.

## Videos (included slow motion)

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/fb35ffa1-e2cc-4949-ac2b-6f3ccacb3470

- slow motion

https://github.com/Expensify/App/assets/56457735/bc20f211-8e8f-4c33-9b31-9b0bc854ede7

</details>

**Comment by garrettmknight** [comment]
@0xmiroslav what do you think of this first proposal?

**Comment by ikevin127** [comment]
@0xmiroslav It's a combination of the BackgroundImage which uses expo-image on iOS and the Lottie sign-in animation which causes this issue. I'd say it's a 70% (BackgroundImage) / 30% (Lottie) in terms of impact on this issue - this can be checked by commenting out either of the 2 and observing how it affects the splash screen hide animation.

From https://github.com/Expensify/App/issues/29067 it looks like the expo-image implementation was chosen as a solution to app performance issues as mentioned in https://github.com/Expensify/App/issues/29067#issuecomment-1827768462:

> Currently, all svg assets in the application are transformed by React-Native-SVG-Transformer into components. This process is very demanding and has an impact on the size of the component tree. Each nested element of the svg resource is transformed into a component. For complex assets such as the BackgroundImage on the SignIn Page, the size of the component tree is tremendous. It has a significant influence on the app performance, because the more components we need to render, the more time it will take.

Since reverting the expo-image PR won't help I think the optimal solution here is to delay the rendering process of the 2 culprits until the splash hide animation is completed.

Here's the test branch: https://github.com/ikevin127/Expensify/tree/fix/34696

**Comment by garrettmknight** [comment]
@aimane-chnaif can you give this one a look at the follow up on this one?

**Comment by ikevin127** [comment]
@aimane-chnaif Logout, then close the app and re-open it and observe the splash hiding while the login page is behind it.

**Comment by ikevin127** [comment]
@aimane-chnaif Its exactly like the last video you posted, that's the issue here.

It only happens on login page when the app is first opened (from kill state, not background) because that's where the BackgroundImage SVG and Lottie animation are rendering when opening the app when the splash is hidding.

You can find more details in my proposal's RCA as to what causes the issue, which is still reproducible on both current staging app and with latest main branch.

**Comment by ikevin127** [comment]
Its exactly like the last video you posted, that's the issue here.

It only happens on login page when the app is first opened (from kill state, not background) because that's where the BackgroundImage SVG and Lottie animation are rendering when opening the app when the splash is hidding.

You can find more details in my proposal's RCA as to what causes the issue, which is still reproducible on both current staging app and with latest main branch.

cc @aimane-chnaif 

Before is latest staging app [v1.4.29-0] on an iPhone 12 mini, after is w/ the fix shared via the test branch.

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/415c7a21-a3cc-4b63-8d80-aa3714c22d6e"> | <video src="https://github.com/Expensify/App/assets/56457735/d0224562-8904-4eb5-b319-e8b64e7afa66"> |

**Comment by ikevin127** [comment]
PR https://github.com/Expensify/App/pull/35185 ready for review!

**Comment by ikevin127** [comment]
Looks like the automation might've failed to set the 7-day hold for payment since the PR https://github.com/Expensify/App/pull/35185#issuecomment-1945392111 was deployed to production ~4 hours ago 🧐

**Comment by garrettmknight** [comment]
Paid!

**Comment by garrettmknight** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ]  [@aimane-chnaif] The PR that introduced the bug has been identified. Link to the PR:
- [ ]  [@aimane-chnaif] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ]  [@aimane-chnaif] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ]  [@aimane-chnaif] Determine if we should create a regression test for this bug.
- [ ]  [@aimane-chnaif] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.
- [ ]  [@garrettmknight] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by garrettmknight** [comment]
@aimane-chnaif can you complete the checklist when you get a chance?
